vantage-agent/lib/charms/vantage_agent/v0/agent_snapper.py

```diff
@@ -0,0 +1,193 @@
+        self._charm.unit.status = WaitingStatus(f"Installing snap for {self._snap_name}")
+        try:
+            self.install_snap()
+        except Exception as e:
+            logger.error(f"## Error installing {self._snap_name}: {e}")
+            self._charm.unit.status = BlockedStatus("Error installing the snap for {self._snap_name}")
+            event.defer()
+            return
+
+        logger.debug(f"Snap for {self._snap_name} installed")
+
+    def handle_config_changed(self, event: ConfigChangedEvent):
+        logger.debug(f"## Processing config changed event for {self._snap_name}.")
+        if not self.is_snap_installed():
+            logger.debug(f"## Snap for {self._snap_name} not installed, deferring event")
+            event.defer()
+            return
+
+        logger.info(f"## Updating config for {self._snap_name}")
+        self.configure_snap()
+
+        if self.is_snap_active():
+            self.run_snap_service("restart")
+
+    def handle_start(self, event: StopEvent):
+        logger.debug(f"## Processing start event for {self._snap_name}.")
+        self.run_snap_service("start")
+        self._charm.unit.status = ActiveStatus()
+
+    def handle_stop(self, event: StopEvent):
+        logger.debug(f"## Processing stop event for {self._snap_name}.")
+        self.run_snap_service("stop")
+        self._charm.unit.status = MaintenanceStatus()
+
+    def handle_remove(self, event: RemoveEvent):
+        logger.debug(f"## Processing remove event for {self._snap_name}.")
+        self.remove_snap()
+
+    ## Operations
+    def is_snap_installed(self) -> bool:
+        cmd = [
+            self._snap_path,
+            "list",
+            self._snap_name,
+        ]
+
+        try:
+            subprocess.check_output(cmd)
+            return True
+        except subprocess.CalledProcessError:
+            return False
+
+    def is_snap_active(self) -> bool:
+        logger.debug(f"### Checking active status for {self._snap_name}.daemon")
+        cmd = [
+            self._snap_path,
+            "services",
+            f"{self._snap_name}.daemon",
+        ]
+
+        try:
+            status_output = subprocess.check_output(cmd).decode("utf-8")
+            status_line = status_output.split("\n")[1]
+            status = status_line.split()[2]
+            return status == "active"
+        except subprocess.CalledProcessError:
+            return False
+
+    def install_snap(self) -> None:
+        channel = self._charm.config["snap-channel"]
+        assert isinstance(channel, str)
+
+        confinement = self._charm.config["snap-confinement"]
+        assert isinstance(confinement, str)
+
+        if not self.is_snap_installed():
+            logger.debug(f"### Installing {self._snap_name}.")
+            self._sys_exec(
+                self._snap_path,
+                "install",
+                "--channel",
+                channel,
+                f"--{confinement}",
```

Comment:
This will fail for non-classic snaps because there's no `--strict` flag for example. I understand that the agents currently use the `classic` confinement but I think it'd be better to make this code agnostic of that in case we decide to change in the future, who knows...

Reply:
I think it will be better to just force `--classic` for now. We can change it in the future if we need to, but I think these snaps will need to have classic confinement no matter what.

Reply:
Resolved in commit e15fece.